jax/experimental/pjit.py

```diff
@@ -0,0 +1,211 @@
+            "All static broadcasted arguments must be passed positionally.")
+      dyn_argnums = [i for i in range(len(args)) if i not in static_argnums]
+      f, args = argnums_partial(f, dyn_argnums, args)
+
+    args_flat, in_tree = tree_flatten(args)
+    for arg in args_flat: _check_arg(arg)
+    flat_fun, out_tree = flatten_fun_nokwargs(f, in_tree)
+
+    in_mesh_axes_flat = tuple(flatten_axes("pjit in_mesh_axes",
+                                           in_tree, in_mesh_axes))
+    out_mesh_axes_thunk = HashableFunction(
+        lambda: tuple(flatten_axes("pjit out_mesh_axes", out_tree(), out_mesh_axes)),
+        closure=out_mesh_axes)
+
+    out = pjit_call_p.bind(
+        flat_fun,
+        *args_flat,
+        in_mesh_axes=in_mesh_axes_flat,
+        out_mesh_axes_thunk=out_mesh_axes_thunk,
+        resource_env=resource_env,
+        name=flat_fun.__name__)
+    return tree_unflatten(out_tree(), out)
+
+  return wrapped
+
+def _pjit_call_impl(fun: lu.WrappedFun, *args, in_mesh_axes,
+                    out_mesh_axes_thunk, resource_env, name):
+  in_avals = [core.raise_to_shaped(core.get_aval(arg)) for arg in args]
+  return _pjit_callable(
+      fun, in_mesh_axes, out_mesh_axes_thunk, resource_env, name,
+      *in_avals)(*args)
+
+def _pjit_translation_rule(c, axis_env, in_nodes, name_stack, backend, name,
+                           call_jaxpr, in_mesh_axes, out_mesh_axes_thunk,
+                           resource_env):
+  # TODO(skye): this should arguably be part of axis_env
+  mesh = resource_env.physical_mesh
+  subc = xc.XlaBuilder(f"pjit_{name}")
+
+  # We assume any extra leading in_nodes are constants and replicate them.
+  num_extra_nodes = len(in_nodes) - len(in_mesh_axes)
+  assert num_extra_nodes >= 0
+  in_mesh_axes = (None,) * num_extra_nodes + in_mesh_axes
+
+  args = []
+  for i, (n, mesh_axes) in enumerate(safe_zip(in_nodes, in_mesh_axes)):
+    # N.B. inlined calls shouldn't have shardings set directly on the inputs or
+    # outputs (set_sharding_proto adds an identity operation).
+    arg = xb.parameter(subc, i, c.GetShape(n))
+    args.append(xb.set_sharding_proto(subc, arg,
+                                      get_sharding_proto(c, n, mesh_axes, mesh)))
+
+  out_nodes = xla.jaxpr_subcomp(
+      subc, call_jaxpr, backend, axis_env, (),
+      extend_name_stack(name_stack, wrap_name(name, "pjit")), *args)
+  out_mesh_axes = out_mesh_axes_thunk()
+  assert len(out_mesh_axes) == len(out_nodes)
+  out_nodes = [xb.set_sharding_proto(subc, out,
+                                     get_sharding_proto(c, n, mesh_axes, mesh))
+               for out, mesh_axes in safe_zip(out_nodes, out_mesh_axes)]
+
+  subc = subc.build(xops.Tuple(subc, out_nodes))
+  return xops.Call(c, subc, list(in_nodes))
+
+pjit_call_p = core.CallPrimitive("pjit_call")
+pjit_call_p.def_impl(_pjit_call_impl)
+xla.call_translations[pjit_call_p] = _pjit_translation_rule
+
+# None indicates replicated axis (partial replication)
```

Comment:
I don't think that this is the same thing as partial replication. You have to have at least one `None` in your axis partitioning to even have a chance of triggering partial replication, but there are partitions that do use Nones and are sharded over the whole mesh. What matters is whether you use all mesh axis names or not!